src/sql-parser/src/parser.rs

```diff
@@ -192,6 +203,193 @@ impl ParserError {
+    }
+
+    pub fn object_type(&self) -> Option<(StatementAction, AuditObjectType)> {
+        self.statement
+            .map(|s| match s {
+                StatementKind::AlterConnection => {
+                    Some((StatementAction::Alter, AuditObjectType::Connection))
+                }
+                StatementKind::AlterIndex => Some((StatementAction::Alter, AuditObjectType::Index)),
+                StatementKind::AlterRole => Some((StatementAction::Alter, AuditObjectType::Role)),
+                StatementKind::AlterSecret => {
+                    Some((StatementAction::Alter, AuditObjectType::Secret))
+                }
+                StatementKind::AlterSink => Some((StatementAction::Alter, AuditObjectType::Sink)),
+                StatementKind::AlterSource => {
+                    Some((StatementAction::Alter, AuditObjectType::Source))
+                }
+                StatementKind::CreateCluster => {
+                    Some((StatementAction::Create, AuditObjectType::Cluster))
+                }
+                StatementKind::CreateClusterReplica => {
+                    Some((StatementAction::Create, AuditObjectType::ClusterReplica))
+                }
+                StatementKind::CreateConnection => {
+                    Some((StatementAction::Create, AuditObjectType::Connection))
+                }
+                StatementKind::CreateDatabase => {
+                    Some((StatementAction::Create, AuditObjectType::Database))
+                }
+                StatementKind::CreateIndex => {
+                    Some((StatementAction::Create, AuditObjectType::Index))
+                }
+                StatementKind::CreateMaterializedView => {
+                    Some((StatementAction::Create, AuditObjectType::MaterializedView))
+                }
+                StatementKind::CreateRole => Some((StatementAction::Create, AuditObjectType::Role)),
+                StatementKind::CreateSchema => {
+                    Some((StatementAction::Create, AuditObjectType::Schema))
+                }
+                StatementKind::CreateSecret => {
+                    Some((StatementAction::Create, AuditObjectType::Secret))
+                }
+                StatementKind::CreateSink => Some((StatementAction::Create, AuditObjectType::Sink)),
+                StatementKind::CreateSource => {
+                    Some((StatementAction::Create, AuditObjectType::Source))
+                }
+                StatementKind::CreateSubsource => {
+                    Some((StatementAction::Create, AuditObjectType::Subsource))
+                }
+                StatementKind::CreateTable => {
+                    Some((StatementAction::Create, AuditObjectType::Table))
+                }
+                StatementKind::CreateView => Some((StatementAction::Create, AuditObjectType::View)),
+                _ => None,
+            })
+            .flatten()
+    }
+}
```

Comment:
We've endeavored to keep the SQL parser as minimal and focused as possible. Since `StatementAction` and the methods on `ParserStatementError` are only necessary for the Segment telemetry, I'd like to pop them out of this crate and make them free functions in (say) src/adapter/telemetry.rs. I'll post a top-level comment about a `TelemeterizedParser` that would be a good home for these methods, in addition to solving some other problems.